src/nupic/encoders/SDRCategoryEncoder.hpp

```diff
@@ -0,0 +1,539 @@
+/* ---------------------------------------------------------------------
+* Numenta Platform for Intelligent Computing (NuPIC)
+* Copyright (C) 2014-2016, Numenta, In  Unless you have an agreement
+* with Numenta, In, for a separate license for this software code, the
+* following terms and conditions apply:
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU Affero Public License version 3 as
+* published by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+* See the GNU Affero Public License for more details.
+*
+* You should have received a copy of the GNU Affero Public License
+* along with this program.  If not, see http://www.gnu.org/licenses.
+*
+* http://numenta.org/licenses/
+* ---------------------------------------------------------------------
+*/
+
+#ifndef NTA_sdrcategorycoder_HPP
+#define NTA_sdrcategorycoder_HPP
+
+#include <iostream>
+#include <string> 
+#include <vector>
+#include <tuple>
+#include <map> 
+
+#include "Encoder.hpp"  
+#include "DecodeResult.hpp"
+#include "nupic/utils/Random.hpp" 
+#include <nupic/types/Types.hpp>
+
+using namespace std;
+
+namespace encoders {
+
+	/**
+	* Encodes a list of discrete categories (described by strings), that aren't
+	* related to each other.
+	* Each  encoding is an SDR in which w out of n bits are turned on.
+	* Unknown categories are encoded as a single
+	* Internally we use a ScalarEncoder with a radius of 1, but since we only encode
+	* integers, we never get mixture outputs.
+	* The SDRCategoryEncoder uses a different method to encode categories
+	*
+	* @numenta.htm.java
+	*/
+	template<typename T>
+	class SDRCategoryEncoder : public Encoder<T> {
```

Comment:
I have an alternative implementation of this class started in PR #278 

It uses a similar but subtly different algorithm.

Reply:
I know. Can you judge which implementation is better, or how much is 278 finished? 
Advantage of this impl is that it imho is quite identical to the python code.

Reply:
Here is a memo I wrote describing the difference between my implementation in PR #278 and the python reference implementation:  https://github.com/htm-community/nupic.cpp/issues/258#issuecomment-466446149

This memo applies to both the RDSE & Category encoder, even though the memo only mentions the RDSE.

Reply:
thanks, I recall the difference. Well..what now? Should we keep both encoders? 

PR278_RDSE:
- no decode -> not an issue, I'd prefer to use classifier anyway. 
- > No strong guarantee that semantically unrelated inputs have a low overlap. This should only be an issue if the encoder is too small or its sparsity is too large. Mitigation: we can quantify these failure conditions and test them with unit-tests.

this could be a problem. I don't have a strong opinion, but given this, I think either both, or this (PR291) RDSE should stay. If people do experiments, it's good to have RDSE/Category implementation matching the numenta's python version. 

PR278_RDSE advantage: 
- speed and memory: I've worked with the RDSE from this PR and speed was not a limiting issue. 
- an advantage of yours approach which I like is that the implementation is simpler.